From fc86a5578afad47cfdab2de24417a6ff3f74bd6c Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentin.casasnovas@vates.tech>
Date: Mon, 11 May 2026 16:22:41 +0200
Subject: [PATCH] !kabi crypto: af_alg - fix use-after-free in af_alg_accept()
 due to bh_lock_sock()

Everything fits so simply hide the change from genksyms.

Fixes: 794706b54241 ("crypto: af_alg - fix use-after-free in af_alg_accept() due to bh_lock_sock()")
Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 crypto/af_alg.c         | 20 ++++++++++++++++++++
 include/crypto/if_alg.h | 11 ++++++++++-
 2 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/crypto/af_alg.c b/crypto/af_alg.c
index 4e5d7d9448f0..e7d4c09ecc1f 100644
--- a/crypto/af_alg.c
+++ b/crypto/af_alg.c
@@ -1205,3 +1205,23 @@ module_init(af_alg_init);
 module_exit(af_alg_exit);
 MODULE_LICENSE("GPL");
 MODULE_ALIAS_NETPROTO(AF_ALG);
+
+#include <linux/build_bug.h>
+
+void kabi_check_struct_alg_sock(void)
+{
+	struct old_alg_sock {
+		/* struct sock must be the first member of struct alg_sock */
+		struct sock sk;
+
+		struct sock *parent;
+
+		unsigned int refcnt;
+		unsigned int nokey_refcnt;
+
+		const struct af_alg_type *type;
+		void *private;
+	};
+	BUILD_BUG_ON(sizeof(struct old_alg_sock) != sizeof(struct alg_sock));
+	BUILD_BUG_ON(offsetof(struct old_alg_sock, type) != offsetof(struct alg_sock, type));
+}
diff --git a/include/crypto/if_alg.h b/include/crypto/if_alg.h
index 11f107df78dc..f96b17862507 100644
--- a/include/crypto/if_alg.h
+++ b/include/crypto/if_alg.h
@@ -34,9 +34,18 @@ struct alg_sock {
 
 	struct sock *parent;
 
+#ifdef __GENKSYMS__
+	/**
+	 * Changed from unsigned int to atomic_t in 794706b54241 ("crypto:
+	 * af_alg - fix use-after-free in af_alg_accept() due to
+	 * bh_lock_sock()") - both are using 4 bytes.
+	 */
+	unsigned int refcnt;
+	unsigned int nokey_refcnt;
+#else
 	atomic_t refcnt;
 	atomic_t nokey_refcnt;
-
+#endif
 	const struct af_alg_type *type;
 	void *private;
 };
